handbook/engineering/https_dev_instance.md

```diff
@@ -0,0 +1,173 @@
+> kill the ngrok proxy when you are done and not let it run for long periods of time.
+
+## Method Two
+
+> A little more involved but works without outside internet.
+
+> NOTE: On a Mac installation this method requires one restart of the machine to bounce the DNS resolving machinery
+> of the operating system. The `/etc/resolver` entry will otherwise be ignored until the next restart. 
+
+This method uses:
+
+* [dnsmasq](http://www.thekelleys.org.uk/dnsmasq/doc.html)
+* [mkcert](https://github.com/FiloSottile/mkcert)
+* [caddy2](https://caddyserver.com/)
+
+### Installing and setting up dnsmasq
+
+* `brew install dnsmasq`
+* Edit `/usr/local/etc/dnsmasq.conf` to add this entry:
+
+```text
+address=/test/127.0.0.1
+```
+
+* Add an entry to `/etc/resolver`:
+
+```shell script
+sudo tee /etc/resolver/test >/dev/null <<EOF
+nameserver 127.0.0.1
+EOF
+``` 
+
+* Add dnsmasq to launchd:
+
+```shell script
+sudo brew services start dnsmasq 
+```
+* If you're on a Mac, restart the machine.
+* Verify that `.test` domains resolve to your local machine:
+
+```shell script
+dig sourcegraph.test @127.0.0.1
+``` 
+
+You should see something like this (look for the answer section):
+
+```text
+; <<>> DiG 9.10.6 <<>> sourcegraph.test @127.0.0.1
+;; global options: +cmd
+;; Got answer:
+;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 15978
+;; flags: qr aa rd ra ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
+
+;; OPT PSEUDOSECTION:
+; EDNS: version: 0, flags:; udp: 4096
+;; QUESTION SECTION:
+;sourcegraph.test.		IN	A
+
+;; ANSWER SECTION:
+sourcegraph.test.	0	IN	A	127.0.0.1
+
+;; Query time: 0 msec
+;; SERVER: 127.0.0.1#53(127.0.0.1)
+;; WHEN: Tue Mar 10 21:05:49 PDT 2020
+;; MSG SIZE  rcvd: 61
+```
+
+* `ping -c 1 this.is.a.test ` should also work:
+
+```text
+PING this.is.a.test (127.0.0.1): 56 data bytes
+64 bytes from 127.0.0.1: icmp_seq=0 ttl=64 time=0.025 ms
+
+--- this.is.a.test ping statistics ---
+1 packets transmitted, 1 packets received, 0.0% packet loss
+round-trip min/avg/max/stddev = 0.025/0.025/0.025/0.000 ms
+``` 
+
+### Installing and setting up mkcert
+
+* `brew install mkcert nss`
+* Generate and install a certificate authority:
+
+```shell script
+mkcert -install
+```
+
+This puts a root certificate and its private key in ` ~/Library/Application\ Support/mkcert`. It also declares
+this root certificate to the system, to Firefox (via nss) and to Java (you can control which certificate stores
+it gets added to with command-line flags to mkcert).
+
+> SECURITY NOTE: Be very careful to protect and not share the root private key. Set up reminders to re-generate
+> this certificate every couple of months. If the private key leaks it gives man in-the-middle powers to an attacker.
+
+* Generate a certificate:
+
+```shell script
+mkcert -cert-file sourcegraph.crt -key-file sourcegraph.key sourcegraph.test
+``` 
+
+The Caddy2 reverse proxy will use this certificate.
+
+### Installing and setting up Caddy2
```

Comment:
Afaik we are already using `nginx` in dev, which means that is already running and everyone already has it installed:

https://github.com/sourcegraph/sourcegraph/blob/c983276311fc4ef4222914bdf6a43b7ce7aa0d56/dev/Procfile#L10

https://github.com/sourcegraph/sourcegraph/blob/master/dev/nginx.conf

Would it be possible to configure nginx in the above conf file to always use HTTPS instead of requiring manual installation of another webserver?

Reply:
it's possible, i'll look into it.

Reply:
ok i looked into it. that nginx serves a purpose (it better be, duh :-)). anyways it's a proxy chain like this: webpack-dev:3080 -> nginx:3081 -> sourcegraph-frontend:3082

i added another server block in the nginx.conf and rerouted the traffic. 

it works. i'm a little bit uncomfortable with it because it very much reminds me of a rube goldberg contraption.

i'll add it to the doc as an option.